File: leetcode/tree/populating_next_right_pointer.go
package main

import "fmt"

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

func connect(root *Node) *Node {
	if root == nil {
		return nil
	}
	var q []*Node
	q = append(q, root)
	for len(q) != 0 {
		size := len(q)
		var pre *Node

		for i := 0; i < size; i++ {
			tmp := q[0]
			q = q[1:]
			if pre != nil {
				pre.Next = tmp
			}
			if tmp.Left != nil {
				q = append(q, tmp.Left)
			}
			if tmp.Right != nil {
				q = append(q, tmp.Right)
			}
			pre = tmp
		}
	}
	return root
}

func main() {
	root := &Node{Val: 1, Left: nil, Right: nil, Next: nil}
	root.Left = &Node{Val: 2, Left: nil, Right: nil, Next: nil}
	root.Right = &Node{Val: 3, Left: nil, Right: nil, Next: nil}
	root.Left.Left = &Node{Val: 4, Left: nil, Right: nil, Next: nil}
	root.Left.Right = &Node{Val: 5, Left: nil, Right: nil, Next: nil}
	root.Right.Left = &Node{Val: 6, Left: nil, Right: nil, Next: nil}
	root.Right.Right = &Node{Val: 7, Left: nil, Right: nil, Next: nil}
	res := connect(root)
	for res != nil {
		curr := res
		for curr != nil {
			fmt.Println(curr.Val)
			curr = curr.Next
		}
		res = res.Left
	}
}

File: leetcode/tree/populating_next_right_pointer.py
from common import Node

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None
        q = []
        q.append(root)
        while len(q) != 0:
            size = len(q)
            prev = None
            for _ in range(size):
                tmp = q.pop(0)
                if prev is not None:
                    prev.next = tmp
                if tmp.left:
                    q.append(tmp.left)
                if tmp.right:
                    q.append(tmp.right)
                prev = tmp
        return root

if __name__ == "__main__":
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    root.right.left = Node(6)
    root.right.right = Node(7)
    res = Solution().connect(root)

    while res:
        curr = res
        while curr:
            print(curr.val)
            curr = curr.next
        res = res.left
